File: frontend/src/types/knowledge.ts
/**
 * 知识库模块类型 — 对齐 backend/app/schemas/knowledge/knowledge_base_schema.py
 */

/** 库级状态（spec §3.1：换模型 reindexing） */
export type KnowledgeBaseStatus = 'ready' | 'reindexing'

/** 检索模式（对齐 backend RetrievalMode StrEnum） */
export type RetrievalMode = 'vector' | 'keyword' | 'hybrid'

/** 库级切块配置（一套，作用于库内所有文档） */
export interface ChunkConfig {
  chunk_size: number
  overlap: number
  strategy: 'recursive'
}

/** KnowledgeBase 实例（GET 返回） */
export interface KnowledgeBase {
  id: string
  name: string
  description: string
  embedding_model_id: string
  embedding_model_name: string
  embedding_dim: number
  chunk_config: ChunkConfig
  status: KnowledgeBaseStatus
  doc_count: number
  chunk_count: number
  retrieval_mode: RetrievalMode
  rerank_model_id: string | null
  rerank_model_name: string | null
  created_at: string
  updated_at: string
}

/** 创建知识库请求体。embedding_dim 由后端随模型锁定，不在此处传。 */
export interface KnowledgeBaseCreatePayload {
  name: string
  description?: string
  embedding_model_id: string
  chunk_config?: Partial<ChunkConfig>
}

/** 更新知识库请求体（部分更新）。不可改 embedding 模型——后端 schema 已锁。 */
export interface KnowledgeBaseUpdatePayload {
  name?: string
  description?: string
  chunk_config?: ChunkConfig
  retrieval_mode?: RetrievalMode
  /** null = 关闭精排；不传 = 不变 */
  rerank_model_id?: string | null
}

// ============================================================================
// Document
// ============================================================================

/** 文档大状态（对齐后端 Document.status） */
export type DocumentStatus = 'pending' | 'processing' | 'completed' | 'failed'

/** 文档细分阶段（对齐后端 Document.stage）
 *
 * - `''` = init 后还没传完
 * - `'uploaded'` = 字节已传完、等向量化
 * - `'queued'` = 已入队、等 worker 取（重试等待期也回落到这里）
 * - `'parsing'` / `'splitting'` / `'embedding'` = 向量化管线进行中（片5）
 */
export type DocumentStage =
  | ''
  | 'uploaded'
  | 'queued'
  | 'parsing'
  | 'splitting'
  | 'embedding'

/** Document 对外形态（对齐后端 DocumentOut） */
export interface Document {
  id: string
  knowledge_base_id: string
  name: string
  file_type: string
  size: number
  char_length: number
  paragraph_count: number
  chunk_count: number
  status: DocumentStatus
  stage: DocumentStage
  error_message: string
  created_at: string
  updated_at: string
}

/** 上传策略（init 返回的判别字段） */
export type UploadStrategy = 'presign' | 'passthrough'

/** init 上传响应（对齐后端 UploadInitOut）
 *
 * - `presign` 时：用 `upload_url` + `headers` 直接 PUT 到 R2
 * - `passthrough` 时：用 `upload_endpoint` multipart POST 给后端
 */
export interface UploadInitOut {
  strategy: UploadStrategy
  document_id: string
  upload_url?: string
  upload_endpoint?: string
  headers: Record<string, string>
}

// ============================================================================
// 检索 / 命中测试
// ============================================================================

/** 单条命中结果（对齐后端 RetrievalHit）
 *
 * - `content`：命中后返回的整段（父子块的父级）
 * - `chunk_text`：实际命中的子块原文，调试切块时看
 * - `title`：段的标题链（`第三章 > 3.1 向量检索`）。与 `doc_name` 合起来构成
 *   完整出处；祖先那几级是 `content` 里没有的信息。无标题区域（txt 全篇、
 *   md 的 frontmatter 与前言）为空串
 * - `score`：相似度 0~1（= 1 - 余弦距离），越大越相关
 * - `matched_by`：hybrid 模式下的命中来源路；单模式检索为 null
 */
export interface RetrievalHit {
  paragraph_id: string
  document_id: string
  doc_name: string
  title: string
  /** 段的起始页；仅 PDF 有，其余格式为 null */
  page: number | null
  content: string
  chunk_text: string
  score: number
  matched_by: 'vector' | 'keyword' | 'both' | null
}

/** 命中测试响应（对齐后端 RetrievalTestOut）：命中列表 + 四段耗时（毫秒）
 *
 * - `embed_ms`：query 向量化耗时（调 embedding 模型，通常占大头）
 * - `search_ms`：检索 SQL 耗时
 * - `rerank_ms`：精排耗时（未开启精排时为 0）
 * - `total_ms`：总耗时
 */
export interface RetrievalTestResult {
  hits: RetrievalHit[]
  embed_ms: number
  search_ms: number
  rerank_ms: number
  total_ms: number
}

// ============================================================================
// Document — 批量操作
// ============================================================================

/** 批量向量化结果（对齐后端 BatchProcessOut）
 *
 * - `triggered`：已入队处理的文档 id（前端乐观标 processing + 轮询）
 * - `skipped`：状态不允许、被跳过的文档 id（前端提示用户）
 */
export interface BatchProcessResult {
  triggered: string[]
  skipped: string[]
}

File: backend/app/models/knowledge.py
"""知识库 / RAG 数据模型。

一簇紧耦合的模型放同一文件：KnowledgeBase / Document / Paragraph / Embedding。
层级：文档 → 段(Paragraph) → 子块（不落表，文本存 Embedding.text）。
"""

from tortoise import fields

from app.db.fields import VectorField
from app.models.base import TimestampMixin, UUIDBaseModel


class KnowledgeBase(UUIDBaseModel, TimestampMixin):
    """知识库：一组文档的集合，建库时锁定一个 embedding 模型。

    - `embedding_model` / `embedding_dim`：锁定的向量化模型与维度（换模型 = 全库重建）
    - `chunk_config`：库级切块配置（chunk_size / overlap / strategy），改了需重切
    - `status`：库级状态（ready / reindexing，给换模型重建用）
    """

    created_by = fields.ForeignKeyField(
        "models.User", related_name="knowledge_bases", on_delete=fields.CASCADE,
    )
    name = fields.CharField(max_length=150, description="库名")
    description = fields.CharField(max_length=500, default="", description="描述")
    embedding_model = fields.ForeignKeyField(
        "models.AIModel", related_name="knowledge_bases", on_delete=fields.RESTRICT,
        description="锁定的 embedding 模型",
    )
    embedding_dim = fields.IntField(description="向量维度，建库时随模型锁定")
    chunk_config = fields.JSONField(default=dict, description="切块配置：chunk_size / overlap / strategy")
    status = fields.CharField(max_length=20, default="ready", description="库级状态：ready / reindexing")

    class Meta:
        table = "knowledge_bases"
        unique_together = (("created_by", "name"),)


class Document(UUIDBaseModel, TimestampMixin):
    """知识库下的文档。

    原文件存对象存储（`storage_key` 指向）；文本解析、切段、切块、向量化后
    分别落到 Paragraph / Embedding。`status` + `stage` + `error_message`
    跟踪异步处理管线。v1 手动向量化：上传只建记录（status=pending）。
    """

    knowledge_base = fields.ForeignKeyField(
        "models.KnowledgeBase", related_name="documents", on_delete=fields.CASCADE,
    )
    name = fields.CharField(max_length=255, description="文档名")
    file_type = fields.CharField(max_length=20, description="源文件类型：md / txt")
    size = fields.IntField(default=0, description="字节数")
    storage_key = fields.CharField(
        max_length=512, description="存储后端的键：S3=对象 key，本地=相对路径",
    )
    char_length = fields.IntField(default=0, description="字符数（冗余展示）")
    paragraph_count = fields.IntField(default=0, description="段数（冗余展示）")
    chunk_count = fields.IntField(default=0, description="子块数 = content 向量数（冗余展示）")
    status = fields.CharField(
        max_length=20, default="pending",
        description="pending / processing / completed / failed",
    )
    stage = fields.CharField(
        max_length=20, default="", description="处理阶段：parsing / splitting / embedding",
    )
    error_message = fields.TextField(default="", description="失败原因")

    class Meta:
        table = "documents"


class Paragraph(UUIDBaseModel, TimestampMixin):
    """段：展示 / 编辑 / 命中后返回给模型的单元（父子块的父级）。

    `content` 是命中后返回给模型的整段文本；段内再切成子块去 embed
    （子块不落表，文本存 Embedding.text）。
    """

    knowledge_base = fields.ForeignKeyField(
        "models.KnowledgeBase", related_name="paragraphs", on_delete=fields.CASCADE,
    )
    document = fields.ForeignKeyField(
        "models.Document", related_name="paragraphs", on_delete=fields.CASCADE,
    )
    content = fields.TextField(description="段全文——命中后返回给模型的就是它")
    title = fields.CharField(max_length=256, default="", description="段 / 章节标题")
    position = fields.IntField(default=0, description="段在文档中的顺序")
    char_length = fields.IntField(default=0, description="字符数")

    class Meta:
        table = "paragraphs"


class Embedding(UUIDBaseModel):
    """向量：embed / 检索单元，一对多指回段（多向量留口子）。

    - 始终挂 `paragraph`：命中后据此返回整段（父子块）。
    - `source_type`：content（v1 唯一）/ question / title。
    - `text`：被 embed 的源文本（content 行 = 子块原文）。
    - `embedding`：pgvector 向量，列不锁维度（兼容多库不同维度）。
    - 不可变派生数据（重嵌入 = 删了重建），故只有 `created_at`、无 `updated_at`。
    """

    knowledge_base = fields.ForeignKeyField(
        "models.KnowledgeBase", related_name="embeddings", on_delete=fields.CASCADE,
    )
    document = fields.ForeignKeyField(
        "models.Document", related_name="embeddings", on_delete=fields.CASCADE,
    )
    paragraph = fields.ForeignKeyField(
        "models.Paragraph", related_name="embeddings", on_delete=fields.CASCADE,
        description="始终有——命中后据此返回整段",
    )
    source_type = fields.CharField(
        max_length=20, default="content", description="content / question / title",
    )
    text = fields.TextField(description="被 embed 的源文本（content 行 = 子块原文）")
    position = fields.IntField(null=True, description="子块在段内顺序（非 content 行可空）")
    embedding = VectorField(description="向量（pgvector，不锁维度）")
    meta = fields.JSONField(default=dict, description="元数据")
    created_at = fields.DatetimeField(auto_now_add=True, description="创建时间")

    class Meta:
        table = "embeddings"
